### DISCUSSION TITLE
Is my understanding of Avalonia MVVM correct?

### DISCUSSION THREAD

**Discussion by al-zakaria** [AvaloniaUI/Avalonia#16103] [Q&A]
I already asked this question, but I read a lot of sites on this topic and wanted to ask if I understood this topic correctly.

this is how I understood MVVM pattern

1 in View you simply write an interface (this is already clear to everyone)
2 in ViewModel I write more. part of the code (you can say almost all the code I have to write in the ViewModel). As I understand it, this should be done for these reasons:
 1. I don’t have to write the name of each UI element, but simply do the data binding to the desired UI element.
 2. Instead of the same names, I use ordinary properties, which are in fact ordinary variables and class fields, which makes the code easier to write and maintain since I again use ordinary variables which are ordinary input or even output data.
 3. I should not write all the functions for the buttons in one class in the Model, which in fact is writing code in a procedural style, which again makes it much more difficult to maintain and change the code, and also prevents normal refactoring. in the MVVM pattern, I can scatter methods across different classes, which makes it easier for me to maintain, change and refactor the code.
 4. this allows you to do normal unit testing, because I again operate with ordinary variables and fields, as well as ordinary input and output data (which I cannot do in code behind, because I cannot check in unit testing to verify that it is written correctly and name of UI elements and all that). look like that's it

3 well, in the Model I write as little code as possible; in fact, this is only the code that I cannot implement in the ViewModel

and there is one more question.
How do you understand what “separation of business logic from View” means?

Unless when I write code in code behind I do not separate the business logic from the View? I don’t write code for the button in the View file.

Sorry for my english

**Comment by LaurentInSeattle** [comment]
Personally, I try my best to place as much code as possible in the model. This code is usually independent from the UI framework and its dependencies can be injected (IoC/DI). This code is then very portable and easily testable. "Fat" view-models do create difficulties for testing and compromise reuse.

**Reply by al-zakaria** [reply]
You didn't understand, on the contrary, I said that "thick" Models and NOT Viewmodels make it difficult to test

**Comment by IanRawley** [comment]
I see it as three parts:
1. Model - the model is responsible for handling raw business data and ensuring it's in a valid configuration.  It's entirely application agnostic, and doesn't particularly care what is attempting to manipulate the data, only that the end result is valid from the model's point of view.
2. View Model - the View Model is responsible for arranging the raw data from the Model is a way that is easy to manipulate by a user, and dictates the kind of manipulations a user can perform in the scope of the objectives of an application.  It's concerned with what it will allow a user to do, but not how exactly the user will trigger those manipulations of data.  It will at times need to request input from the user, but it isn't concerned about who the user is, where they are, or how exactly they are asked for the input or provided it, only that they did or did not provide what was requested.
3. View - the View takes what the View Model will allow a user to do, and provides the exact means for the user to do that.  Be that a command line prompt, buttons and text boxes, splitting into dialogs, etc.  When the View Model indicates it needs input, it's the View's job to present that request to the user and collect it.

While each level is aware of the levels beneath it, they are completely unaware of what is above.  Which is to say the Model knows only of the Model, the View Model knows the details of both the Model and View Models, and the View knows of all 3 areas.  As a result, each level above the Model can be entirely replaced without needing to modify the levels under it.  A set of View Models for a data entry app can be reused in a read only data viewing app by replacing all the views with a different set that don't provide any way for a user to change data and don't present options the View Model may provide that don't make sense in the new context.  A GUI app can be built to replace a CLI app by building a GUI on top of the existing tested and hopefully proven View Models, reducing the need to re-implement and re-test.

**Reply by al-zakaria** [reply]
but in theory this is what I said? and it turns out that my understanding of MVVM in Avalonia is correct?

**Comment by thevortexcloud** [comment]
> I don’t have to write the name of each UI element, but simply do the data binding to the desired UI element.

This is not an MVVM specific benefit. Rather just a feature most XAML frameworks have. It's possible to use MVVM without any bindings or XAML, it's just rather painful in most cases.

> Instead of the same names, I use ordinary properties

Sort of. In C#/XAML specifically you would use properties that raise change notifications via `INotifyPropertyChange`. The view will automatically listen for the notifications and know when to refresh the display of them. But apart from that, yes they are normal properties. If you don't need the view to update/bindings you can skip raising the change notifications entirely.

> I should not write all the functions for the buttons in one class in the Model

Your model should not contain any event handlers for buttons. That's what the code behind for the view is for. You can put all your button handler specific events in there if it makes sense to do so.

> in the MVVM pattern, I can scatter methods across different classes, which makes it easier for me to maintain, change and refactor the code.

MVVM is primarily concerned with separation of concerns in a specific way that decouples the presentation logic from the application logic. Beyond that it really does not matter that much how you structure your code. Separating across multiple classes is also not unique to MVVM, many other patterns use it. EG MVC, MVU, etc. 

But as a general rule you should avoid making god classes. 

> his allows you to do normal unit testing, because I again operate with ordinary variables and fields, 

MVVM itself does not allow for unit testing. You need to write code that is testable (EG using DI, abstractions, single responsibility classes, highly cohesive classes, etc). It does however make it easier to test by making sure all your presentation logic can be ignored and your main application logic can be tested without it.

>  in the Model I write as little code as possible;

Depending on what you are doing, this is possibly very wrong. The model should contain your main application logic. The view model should wrap it/call it. EG If you had a permissions system I would expect:

1. The model to determine who has permission to do what, and also update what ever the data storage is for the permissions
2. The view model to request information from the model about permissions, and also _notify_ the model when updates are required. The view model can expose other view models, such as permission information to the view. Although it's not unusual data to only flow in a single direction after initialisation.
3. The view to display the exposed information from the view model

> How do you understand what “separation of business logic from View” means?

It means your view should not be responsible for all of your app's logic. For example in a classic winforms app, most people shove all the logic into the code behind file. To reuse the the above example: The view in the winforms app would be directly responsible for querying the data store, presenting it to the user and doing pretty much everything. In MVVM the view is only concerned with what/how it is presenting data. Rather than how it got the data and what gets done with it.

**Reply by al-zakaria** [reply]
Now I'm even more confused. I thought that the Model is a .axaml.cs file and the Model is a separate file. why is all this needed? I don't understand at all. then why do we need ViewMdel and Model, and why do we need .axaml.cs files? I'm completely confused. How are they all different then? How then to make Model files? will they be regular .cs files or what?

**Reply by IanRawley** [reply]
MVVM isn't particularly concerned with file arrangement, but the classes within. 

.axaml files are your XAML view files, the corresponding .axaml.cs files are the code-behind files for your views.  .axaml.cs files are where your Button click event handlers and things specific to user interaction will go.

Your Model classes and ViewModel classes will go in .cs files, because they're pure C#, assuming that's the language you're using.  Ideally (I feel) the 3 will be split into separate projects, to help make them more easily reusable and ensure each is only dependent on libraries that they actually need.  For instance as the ViewModel and View projects have no interest in the Views, there's no reason for them to depend on Avalonia packages.

Avalonia is a UI framework, which means it really only deals with the View level.  ViewModels and Models are outside its scope, even though the Avalonia templates do provide ViewModel and Model directories in the projects it produces.

**Reply by thevortexcloud** [reply]
> I thought that the Model is a .axaml.cs file

Those are the code behind files, they are part of the view and let you do many things you can't do with XAML. The models are elsewhere and are typically POCO. You don't have to put them in different projects if you don't want to, it's really up to you how you structure your projects as long as it follows MVVM principles.

**Reply by al-zakaria** [reply]
I still don’t understand why all this is needed. It turns out that there are 4 types of files. this is View, ViewModel, Model and code behind is now the question. Why should I make button handlers in code behind if I can do it in ViewModel? the only thing I couldn't do through the ViewModel was adding new elements to the view. and what is the point of Model if there is a ViewModel. I write as much code as possible in the ViewModel and only a small part of the code in the code behind (.axaml.cs files). and I have never even used Model because there is a ViewModel and code behind. is this acceptable? and is my understanding of MVVM even half correct?

**Reply by timunie** [reply]
well... it depends on the size of your App and what you need to do. I personally have most of the stuff in my ViewModel and View as well. Where possible, I use Commands and Bindings, no code behind. I hope this tutorial may help you: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/CompleteApps/SimpleToDoList

**Reply by thevortexcloud** [reply]
> I still don’t understand why all this is needed

MVVM exists to solve a specific set of problems. The main ones being:

* Ensuring code portability by not coupling application logic to the presentation logic
* Making code reusable, and modular by clearly defining the roles of each component

> It turns out that there are 4 types of files. this is View, ViewModel, Model and code behind is now the question

Technically, the code behind is _part of_ the view. The XAML compiler turns the XAML file into a .NET object which gets automagically merged with the code behind. EG if you make a named button in XAML you can reference the name in the code behind file via a strong typed reference to that specific control instance. But yes, the code behind and XAML files are semantically different things.

> hy should I make button handlers in code behind if I can do it in ViewModel?

Technically, you can't do a button handler in the view model. The view model can expose a _command_ which the view can execute when the button is pressed, but this is different from an event handler. The command is abstraction which has no dependency on the view, as such it can also be used without the view at all or without the view model caring about how it was executed. EG it could trigger when a button is pressed, or when the user double clicks on something. The view model does not really care. The view model does not and should not know things like buttons exist.

The view however does care about things like which command gets fired and _how_ it gets fired. For example, many controls in Avalonia have a double click event. It is not possible to bind a command directly to the event. Instead you can subscribe to the event in the code behind file for the specific control which is being clicked, and then execute the command on the view model. 

>  and I have never even used Model because there is a ViewModel and code behind.

Depending on what you are doing, this is possibly fine. For more complex apps however, I would guess you are doing something wrong if this is the case. 

**The view model is primarily concerned with getting data ready for presentation, without presenting it itself.**

The model is responsible for everything the view and view model does not do. The model could be as simple as a class that writes and reads a text file, to something much more complicated like a mathematical expression parser. In either case, the view model would interact with the model classes to expose their functionality to the view (EG via commands which can be bound to buttons). In this way:

* The model knows nothing about the view model and can be used completely independently of the view and view model
* The view model knows about the model and has logic to interact with it and expose it in a way that can present the output/handle input to it
* The view handles direct user interactions with the application and relays things as needed to the the view model.

With all that said, most people are fairly lazy and don't bother to strictly follow MVVM, as it can be fairly annoying to do right. A lot of people just put _all_ their main application logic in their view models. Even if that's not correct.

Also if it helps, here is a simple metaphor:

Imagine a water bottle. The view is the label, the view model is the bottle itself, and the model is the water. The water is what does  all the interesting stuff but without the bottle you only have a puddle, and without the label nobody knows what's in the bottle.

**Reply by al-zakaria** [reply]
I finally figured it out!! Thanks a lot !!!
can I ask one more question? so that I can understand that I understood everything correctly?
1. Is the model a regular C# file, the same as the ViewModel?

2. For example, I am writing some kind of application that has a connection to a database (for example, my favorite MySQL), and also parses some sites (for example, a site with football matches) and should all this be included in the model? and class objects and calls to functions of these classes should be carried out in the ViewModel? that is, for example, in Model I create a class with a connection to the database and a class with which the site is parsed with the time of football matches. then I make functions (which will be used as commands in the View. For example, using the RelayCommand decorator), in which I call objects of classes and functions of these classes. then in the View I call a command from the ViewModel, which in turn calls large and complex methods from the Model. in fact, it turns out that the ViewModel is a set of commands. that is, the ViewModel consists of properties and methods that simply call commands from the Model, and theoretically the whole point of methods in a ViewModel is to contain simple code and mostly just call complex methods from the Model.

and we can also say that each MVVM element knows about itself and what is below it, but does not know what is above it.
for example an example with the same bottle
label(View) knows about itself, about ViewModel and about Model
the bottle itself (ViewModel) knows about itself and the Model
water (Model) knows only about itself
I hope I said it right.
in this case, the exception to the rule is the code behind, because it knows about both itself and the View (the one above it), and as I learned from the good code behind, it is better not to use it at all.
Sorry for saying so much, but I hope now I said everything correctly

and as I understand it, the division into ViewModel and Model is precisely the separation of business logic from the view and allows you to test the Model and easily transfer the code since the Model is not tied to the View

**Reply by thevortexcloud** [reply]
> Is the model a regular C# file, the same as the ViewModel?

Generally, yes. It can technically be something else though, but that's a more esoteric situation.

> and should all this be included in the model

Yes. Your data objects would be their own separate classes, and then typically you would make a dedicated view model that wraps the dataclass. EG something like

Model data class
```csharp
public class MyWebSiteInfo {
public string WebSiteName { get; set; }
}
```
View model class:
```csharp
public class MyWebSiteInfoViewModel {
private readonly MyWebSiteInfo _info;

public string WebSiteName =>_info.WebSiteName;

public class MyWebSiteInfoViewModel(MyWebSiteInfo info) {
_info = info;
}
}
```
Application data access model class:

```csharp
public class MyWebSiteParser {
public MyWebSiteInfo GetInfo() {
}
}
```

Then you might have some other view model that calls `MyWebSiteParser.GetInfo()` and then passes the result into the constructor of `MyWebSiteInfoViewModel`, which the secondary view model then uses the newly instantiated `MyWebSiteInfoViewModel` by assigning it to a property that the view can bind to.

>  and theoretically the whole point of methods in a ViewModel is to contain simple code and mostly just call complex methods from the Model.

This is correct but the reasoning is wrong. The view model is halfway between the view and the model, it's not really meant to do anything itself beyond interact with the view model/indirectly handle data from the view to then interact with the model. The complexity of the method is not really relevant. but as I said most people are fairly lazy and inevitably put things in the view model they are not technically meant to do.

> simply call commands from the Model

You would not have an `ICommand` in the model. You would however have methods though that the view model can call. A command would only ever be on the view model, since it needs to be accessed by the view.

> and we can also say that each MVVM element knows about itself and what is below it, but does not know what is above it.
> for example an example with the same bottle
> label(View) knows about itself, about ViewModel and about Model
> the bottle itself (ViewModel) knows about itself and the Model
> water (Model) knows only about itself
> I hope I said it right.

This is almost right. The label can see the bottle since it sits on top of the bottle, but it can't see the water since the bottle is in the way. If it wants to get to the water, it needs to ask the the view model to do something with it (EG drink it).

> and as I learned from the good code behind, it is better not to use it at all.

You should avoid it if you can since it can make your app harder to unit test and less portable. But it can't always be avoided, and in many cases it makes far more sense to use the code behind than contort your way into using the view model. Especially when you are relying on functionality provided by the view.

> the exception to the rule is the code behind, because it knows about both itself and the View 

As I said, the code behind is the view. It's just C# instead of XAML. Anything you can do in XAML, you can do in code behind. You can even access the bound view model from it by using the `DataContext` property in the code behind. However you can't do everything you can do in code behind with XAML. Most people use a mixture of both as it's the only way to achieve the desired view results. EG in Avalonia, only the view can display file dialogues.

https://docs.avaloniaui.net/docs/basics/user-interface/file-dialogs

As such the view (code behind) is the only place you can actually directly prompt for a file. The view model can't do it since that would make it aware of how the view works. There are indirect work arounds that don't break MVVM (EG ReactiveUI interactions), but that's a different topic.

> and as I understand it, the division into ViewModel and Model is precisely the separation of business logic from the view and allows you to test the Model and easily transfer the code since the Model is not tied to the View

This sounds correct. But as I mentioned before, you need to write code that is testable in the first place. Using MVVM does not guarantee that. But that's a whole different discussion.
